pageserver/src/error.rs

```diff
@@ -0,0 +1,61 @@
+//!
+//! The `error` module provides a general [`Error`] struct representing different types of
+//! encountered errors when running the page server.
+//!
+
+use thiserror::Error;
+
+#[derive(Error)]
+pub enum Error {
+    #[error("Postgres IO error: {0:?}")]
+    PostgresIO(#[from] PostgresIOError),
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+#[derive(Error)]
+pub enum PostgresIOError {
+    #[error("Connection reset by peer")]
+    ConnectionReset,
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+// # Debug formatting for error structs
+//
+// We need to implement the `fmt::Debug` trait manually instead of deriving the trait
+// because the `{error:?}` format is used when logging the application's errors and
+// we want to have different formats for different types of encountered errors.
+//
+// - For internal/unexpected error, which is parameterized by `anyhow::Error`,
+// print the source error in the debug format, which also includes the error's backtrace [1].
+//
+// - For expected error, print the error message (using the `{error:#}` format).
+//
+// [1]: Note this only applies when the "backtrace" feature is enabled for the `anyhow` crate.
+
+impl std::fmt::Debug for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Internal(err) => {
+                write!(f, "internal error {err:?}")
+            }
+            _ => {
+                write!(f, "{self:#}")
+            }
+        }
+    }
+}
```

Comment:
I think this is too high level to make the decision of what to print. In a different context, losing the connection might be very unexpected and would warrant a stacktrace. I think it'd be better to check for the ConnectionReset error somewhere in `page_service_conn_main`, and have it return with Ok, instead of propagating the error to the generic error handler in `thread_wrapper`.

Reply:
So yeah, the api user is able to choose himself what to write `{e:?}` or `{e:#}`, we don't want to hide the possibility of doing both, as it's now. 
We can follow anyhow's pattern and print both errors the same way.

Reply:
>  In a different context, losing the connection might be very unexpected and would warrant a stacktrace. I think it'd be better to check for the ConnectionReset error somewhere in page_service_conn_main, and have it return with Ok, instead of propagating the error to the generic error handler in thread_wrapper.

Do you mean that losing PG client connection in `page_service_conn_main` is expected, but in other cases, it might be unexpected? 

Note that the `PostgresIOError::ConnectionReset` is only constructed inside the `page_service_conn_main` by looking into the type of error returned by `PostgresBackend::run` function. So if we want to handle this connection error separately and return Ok, constructing this error is redundant.

Reply:
> So yeah, the api user is able to choose himself what to write `{e:?}` or `{e:#}`, we don't want to hide the possibility of doing both, as it's now. 

Note that the printing backtrace is only applied to `anyhow::Error`, e.g if we debug print the `PostgresIOError::ConnectionReset`, it will output ` PostgresIO(ConnectionReset)`

Reply:
After looking into the code again, I realize that instead of manually implementing the `Debug` trait for the page server errors, I can just modify the how the error is formatted with `fmt::Display`  and update the log codes to format the error using the pretty-print `{:#}` format instead of the debug `{:?}` format.

Changes are implemented in https://github.com/neondatabase/neon/pull/1713/commits/965228e44469ac5e90a55db72a7d04c315a0d088